Assets/Scripts/NPCs/StateMachine/States/Wander.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class Wander : State
+{
+    private Vector3 _target;
+    private bool reachedTarget = true;
+    
+    /*
+     * After a random number of moves (between 1 and 10), the NPC stop moving for few seconds.
+     */
+    private int numberOfMoves = 0;
+    private float waitTimeStamp = 0;
+    
+    public Wander()
+    {
+        name = "Wander";
+        numberOfMoves = (int)Random.Range(1.0f, 10.0f);
+    }
+
+    public override void Move(GameObject player, GameObject npc)
+    {
+        if (waitTimeStamp < Time.time)
+        {
+            Wandering(npc);
+        }
+    }
+    
+    private void Wandering(GameObject npc)
+    {
+        if (reachedTarget)
+        {
+           FindNewTarget(npc);
+        }
+        
+        MoveTowardsTarget(npc);
+
+        if (Vector3.Distance(npc.transform.position, _target) < 1.0f)
+        {
+            reachedTarget = true;
+            numberOfMoves -= 1;
+        }
+
+        if (numberOfMoves == 0)
+        {
+            StopMoving(npc);
+        }
+    }
+
+    private void FindNewTarget(GameObject npc)
+    {
+        Vector3 npcVelocity = npc.GetComponent<NPCManager>().Velocity;
```

Comment:
I wonder if we could make move all the GetComponent() calls outside of this class and store the values in variables, since Rider seem to think that they are expensive method evocation. I saw that in NPCStateMachine we are already storing a NPCManager, so maybe instead of parsing the npc here we could make the GetComponent call in NPCManager and pass the velocity as an argument?

Reply:
Yeah, I forgot to move them into variables when I was cleaning the code. Will do!